hera/nfs/ci_workflow.py

```diff
@@ -0,0 +1,78 @@
+                    requests={
+                        "storage": m.Quantity(__root__="1Gi"),
+                    }
+                ),
+            ),
+        )
+    ],
+        # The container build step copies the code from the NFS share to this ephemeral volume, and then runs the build.
+        # We do this a) because NFS shares are slow and b) to allow us to run another workflow step alongside the build if we wish.
+    volumes=[EmptyDirVolume(name="container-build")],
+    arguments=[
+        Parameter(name="app_repo", value=""),
+        Parameter(name="git_branch", value=""),
+        Parameter(name="target_branch", value=""),
+        Parameter(name="container_tag", value=""),
+        Parameter(name="container_image", value=""),
+        Parameter(name="dockerfile", value=""),
+        Parameter(name="path", value=""),
+    ],
+    annotations={
+        'workflows.argoproj.io/description': '''A basic CI leveraging Argo Workflows.
+
+The Workflow...
+
+* pulls a repo from git. Specifically pulling a branch based on a pull request;
+* merges the target branch into it;
+* modifies the html that will be copied into the container to inject the unique name of the running workflow;
+* builds a container from a Dockerfile and pushes to a registry;
+* deploys an Argo CD application that uses the newly-built container to deploy a static website.
+
+It does not pretend to be a definitive example, but it aims to inspire. In order to make this a semi-usable example, we have cut a number of security corners. Please don't just blindly run this in production.
+''',
+        'workflows.argoproj.io/maintainer': 'Pipekit Inc',
+        'workflows.argoproj.io/maintainer_url': 'https://github.com/pipekit/argo-workflows-ci-example',
+        'workflows.argoproj.io/version': '>= 3.3.6',
+    },
+) as ci_workflow:
+    delete_application_resource = Resource(name="delete-application",
+                  action="delete",
+                  manifest=deploy_application_manifest
+                  )
+    with DAG(name="main") as main:
+        git_checkout_task = Task(name="git-checkout", template_ref=m.TemplateRef(
+            name=git_checkout_pr.name, template=git_checkout_pr.entrypoint))
+        html_modifier_task = Task(name="html-modifier", template_ref=m.TemplateRef(
+            name=html_modifier.name, template=html_modifier.entrypoint))
+        container_build_task = Task(name="container-build", template_ref=m.TemplateRef(
+            name=container_build.name, template=container_build.entrypoint))
+        deploy_application_task = Task(name="deploy-application", template_ref=m.TemplateRef(
+            name=deploy_application.name, template=deploy_application.entrypoint))
+        delete_application_task = delete_application_resource(name="delete-application", depends="deploy-application.Failed")
+        git_checkout_task >> html_modifier_task >> container_build_task >> deploy_application_task
```

Comment:
This works for users who know they can rely on templates + know what templates they should reference in the first place. If this is what the client wants, then this 100% works! 

An alternate version is to expose the jobs directly rather than the templates. For example, if `git_checkout_pr` exposes [this container](https://github.com/pipekit/argo-workflows-ci-example/pull/55/files#diff-1d71acae98fabc89b77a39447393b290870cb885ea726215634440eb82bbf2e6R37), and similarly all the other containers/resources are exposed by the other modules, then this DAG can be expressed as:

```
with DAG(name="main"):
    gco = git_checkout_cont(name="git-checkout")
    html_modify = html_modifier_cont(name="html-modifier")
    ...
    deploy = deploy_cont(name="...")
    gco >> html_modify >> ... >> deploy
```

The above is more "Pythonic". Hera made those independent objects (resource, containers, steps, dag, etc, etc)  callables so that they feel more like Python. You can call them, and pass the arguments of task/step depending on the context the functions are called in. For instance, if you call a container inside a `DAG` it gets transformed into a `Task` automatically upon calling. Similarly, it gets turned into a `Step` if it's inside a `Steps` context.

Reply:
^ goal is to increase component reusability (multiple jobs can reuse containers, resources, etc) and reduce the need to rely on exploring AW to know what workflow templates are available in the first place (since you need to know what template refs to use)

Reply:
I've changed it to your suggestion here. The DAG doesn't match the yaml any more though, but I was aiming for more Pythonic.